Lib/test/test_typing.py

```diff
@@ -2654,6 +2654,78 @@ class D(PNonCall): ...
         with self.assertRaises(TypeError):
             issubclass(D, PNonCall)
 
+    def test_no_weird_caching_with_issubclass_after_isinstance(self):
+        @runtime_checkable
+        class Spam(Protocol):
+            x: int
+
+        class Eggs:
+            def __init__(self) -> None:
+                self.x = 42
+
+        # gh-104555: ABCMeta might cache the result of this isinstance check
+        # if we called super().__instancecheck__ in the wrong place
+        # in _ProtocolMeta.__instancecheck__...
+        self.assertIsInstance(Eggs(), Spam)
+
+        # ...and if it did, then TypeError wouldn't be raised here!
+        with self.assertRaises(TypeError):
+            issubclass(Eggs, Spam)
+
+    def test_no_weird_caching_with_issubclass_after_isinstance2(self):
+        @runtime_checkable
+        class Spam(Protocol):
+            x: int
+
+        class Eggs: ...
+
+        # gh-104555: ABCMeta might cache the result of this isinstance check
+        # if we called super().__instancecheck__ in the wrong place
+        # in _ProtocolMeta.__instancecheck__...
+        self.assertNotIsInstance(Eggs(), Spam)
+
+        # ...and if it did, then TypeError wouldn't be raised here!
+        with self.assertRaises(TypeError):
+            issubclass(Eggs, Spam)
```

Comment:
This test fails on 3.11. We could _consider_ backporting a version of this PR to 3.11, but I'm not sure if that would be a good idea or not. It's a somewhat invasive bugfix.